### PR TITLE
openocd-adapters: select transport based on adapter

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Things were a bit muddled since some boards come with a debug adapter while other boards have to use an external one.

This allows for more flexibility in external debug adapter use. The Adapter will select it's default transport, but boards can overwrite them if they need something special (this can only be JTAG instead of SWD since `hla_swd` is an implementation detail of the adpater and should not be selected by the board.

There are some boards that don't use the `DEBUG_ADAPTER` property at all and define everything in their private `openocd.cfg`.

Those can of course still select their own transport, but ideally they should be transferred to `openocd-adapters/` so the adapters can be shared with other boards.

### Testing procedure

Flashing should still work.
I tried

 - [x] flashing a Kinetis board with OpenOCD
 - [x] flashing a nRF52 board with OpenOCD
 - [x] flashing a board that uses `iotlab.inc.mk`
 - [x] flashing a board that uses JLink with OpenOCD (all the nrf52xxxdk boards come with an on-board JLink)
 - [ ] flashing a mulle board with OpenOCD

### Issues/PRs references

alternative to #13858

### CODE REVIEW

**Comment by aabadie** [comment]
> Flashing should still work.

Can you provide a list of board/adapter to test (stlink, nrf, kinetis, etc) ? This will simplify the tracking of what needs to be tested.

> alternative to #13858

I also prefer this approach over #13858 as it's more generic and covers all openocd adapters.

**Comment by benpicco** [comment]
> Can you provide a list of board/adapter to test (stlink, nrf, kinetis, etc) ? This will simplify the tracking of what needs to be tested.

Sure! Well for most adapters, this basically just shuffles around the location of `-c 'transport select swd'` inside `OPENOCD_ADAPTER_INIT` so I don't think we need to test them all.

What should be tested is

 - [ ] flashing a Kinetis board with OpenOCD
 - [ ] flashing a nRF board with OpenOCD
 - [ ] flashing a board that uses `iotlab.inc.mk`
 - [ ] flashing a board that uses JLink with OpenOCD (all the nrf52dk boards come with an on-board JLink)
 - [ ] flashing a mulle board with OpenOCD

**Comment by gschorcht** [comment]
> I tried with a nRF52810 board and one of the common ST-Link v2 dongles and wasn't successful, but OpenOCD got called with the right parameters so (`hla_swd`), it might well be that OpenOCD does not support this device yet.

It should work because all Nucleo boards are flashed with OpenOCD, ST-LINKV2 and `hla_swd` as transport protocol.
I have tested it with the Waveshare nRF52840 Eval Kit and it also works with the following flashing command :+1:
```
PROGRAMMER=openocd DEBUG_ADAPTER=stlink make BOARD=waveshare-nrf52840-eval-kit -C tests/shell flash
```

**Comment by gschorcht** [comment]
@benpicco Since this PR also solves the problem addressed in PR #13858, I will close PR #13858. However, I would like at least the documentation to be expanded, as suggested in the [PR #13858](https://github.com/RIOT-OS/RIOT/pull/13858/files#diff-67f72f99564bc2d3221028153bd7dc2f). This could be done piggybacked with this PR.

**Comment by gschorcht** [comment]
Looks good to me and makes sense. Unfortunately I can't support the test because I don't have the boards mentioned in the test list :worried:

**Comment by benpicco** [comment]
Flashing the `openlabs-kw41z-mini` with OpenOCD also still works.

File: makefiles/tools/openocd-adapters/stlink.inc.mk [hunk 1]
```diff
@@ -3,15 +3,16 @@
 STLINK_VERSION ?= 2-1
 
 # Use STLINK_VERSION to select which stlink version is used
-OPENOCD_ADAPTER_INIT ?= \
-  -c 'source [find interface/stlink-v$(STLINK_VERSION).cfg]' \
-  -c 'transport select hla_swd'
+OPENOCD_ADAPTER_INIT ?= -c 'source [find interface/stlink-v$(STLINK_VERSION).cfg]'
+
+# ST-Link only supports hla_swd
+OPENOCD_TRANSPORT = hla_swd
```
Comment (maribu) [inline makefiles/tools/openocd-adapters/stlink.inc.mk:9]: [The documentation](http://openocd.org/doc/html/Debug-Adapter-Configuration.html#SWD-Transport) says the [newer ST-Link (v2 and v3)](http://openocd.org/doc/html/Debug-Adapter-Configuration.html#st_005flink_005fdap_005finterface) support in addition `dapdirect_swd`, which seems to be preferred over `hla_swd` when available.

Comment (benpicco) [reply makefiles/tools/openocd-adapters/stlink.inc.mk:9]: Hm is there some way to detect if this is supported?

I've updated the ST-LINK on my `stm32f429i-disco` to V2J37S0 and OpenOCD gives me

```
Open On-Chip Debugger 0.10.0+dev-01363-g480ba8ca8-dirty (2020-07-27-13:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2.cfg is deprecated, please switch to interface/stlink.cfg
Error: Debug adapter doesn't support 'dapdirect_swd' transport
```

Comment (maribu) [reply makefiles/tools/openocd-adapters/stlink.inc.mk:9]: My point wasn't really to handle this, but to update the comment and use `?=` to allow users to use `dapdirect_swd`, if they are using a recent ST-Link with a recent OpenOCD.

```Makefile
# All ST-Link adapters support hla_swd, so use that for simplicity
OPENOCD_TRANSPORT ?= hla_swd
```

Comment (benpicco) [reply makefiles/tools/openocd-adapters/stlink.inc.mk]: Alright!
The idea was that a board can set `OPENOCD_TRANSPORT = swd`, so we'll need to prefix that if `OPENOCD_TRANSPORT` is already set.

**Comment by fjmolinas** [comment]
Test results on `IoT-LAB` flashing still works, we can skip `mulle` IMO, not widely available, can't be bought, I think there is one in FU-BERLIN, but that is it AFAIK.

```
IOTLAB_NODE=m3-9.saclay.iot-lab.info BOARD=iotlab-m3 make -C examples/default/ flash term
```

```
iotlab-node --jmespath='keys(@)[0]' --format='int'  --list saclay,m3,9 --update /home/francisco/workspace/RIOT3/examples/default/bin/iotlab-m3/default.elf | grep 0
0
ssh -t molina@saclay.iot-lab.info 'socat - tcp:m3-9.saclay.iot-lab.info:20000'
help
help
Command              Description
---------------------------------------
reboot               Reboot the node
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
random_init          initializes the PRNG
random_get           returns 32 bit of pseudo randomness
rtc                  control RTC peripheral interface
ifconfig             Configure network interfaces
txtsnd               Sends a custom string as is over the link layer
saul                 interact with sensors and actuators using SAUL
>
```

**Comment by leandrolanzieri** [comment]
> Test results on `IoT-LAB` flashing still works

I'm not sure testing like this is correct in this case. On current master I'm getting the following error when trying to flash the `iotlab-m3`:
```
Info : FTDI SWD mode enabled
swd
Warn : Interface already configured, ignoring
trst_and_srst separate srst_nogate trst_push_pull srst_open_drain connect_assert_srst

Error: SWD mode is active but SWD_EN signal is not defined
```
There seems to be no definition of `SWD_EN` in [the configuration file](https://github.com/analogdevicesinc/openocd/blob/master/tcl/interface/ftdi/iotlab-usb.cfg).

**Comment by fjmolinas** [comment]
> I'm not sure testing like this is correct in this case. On current master I'm getting the following error when trying to flash the iotlab-m3:

You are absolutely right, iot-lab uses its own flashing scheme. Sorry for letting this slip.

**Pull request merged**